src/protocols/secure_channel/PASESession.cpp

```diff
@@ -346,133 +349,200 @@ CHIP_ERROR PASESession::DeriveSecureSession(SecureSession & session, SecureSessi
-    size_t resplen = kPBKDFParamRandomNumberSize + sizeof(uint64_t) + sizeof(uint32_t) + mSaltLength;
+    uint8_t responderRandom[kPBKDFParamRandomNumberSize] = { 0 };
+    ReturnErrorOnFailure(DRBG_get_bytes(responderRandom, kPBKDFParamRandomNumberSize));
 
-    size_t sizeof_point = sizeof(mPoint);
-
-    uint8_t * msg = nullptr;
-
-    resp = System::PacketBufferHandle::New(resplen);
+    size_t data_len = EstimateTLVStructOverhead(
+        kPBKDFParamRandomNumberSize + kPBKDFParamRandomNumberSize + sizeof(uint16_t) + sizeof(uint16_t) + sizeof(uint8_t), 5);
+    System::PacketBufferHandle resp = System::PacketBufferHandle::New(data_len);
     VerifyOrReturnError(!resp.IsNull(), CHIP_ERROR_NO_MEMORY);
 
-    msg = resp->Start();
+    System::PacketBufferTLVWriter tlvWriter;
+    tlvWriter.Init(std::move(resp));
 
-    // Fill in the random value
-    ReturnErrorOnFailure(DRBG_get_bytes(msg, kPBKDFParamRandomNumberSize));
+    TLV::TLVType outerContainerType = TLV::kTLVType_NotSpecified;
+    ReturnErrorOnFailure(tlvWriter.StartContainer(TLV::AnonymousTag, TLV::kTLVType_Structure, outerContainerType));
+    ReturnErrorOnFailure(tlvWriter.Put(TLV::ContextTag(1), initiatorRandom));
+    ReturnErrorOnFailure(tlvWriter.PutBytes(TLV::ContextTag(2), responderRandom, sizeof(responderRandom)));
+    ReturnErrorOnFailure(tlvWriter.Put(TLV::ContextTag(3), GetLocalKeyId(), true));
 
-    // Let's construct the rest of the message using BufferWriter
+    if (!initiatorHasPBKDFParams)
     {
-        Encoding::LittleEndian::BufferWriter bbuf(&msg[kPBKDFParamRandomNumberSize], resplen - kPBKDFParamRandomNumberSize);
-        bbuf.Put64(mIterationCount);
-        bbuf.Put32(mSaltLength);
-        bbuf.Put(mSalt, mSaltLength);
-        VerifyOrReturnError(bbuf.Fit(), CHIP_ERROR_NO_MEMORY);
+        TLV::TLVType outerContainer;
+        ReturnErrorOnFailure(tlvWriter.StartContainer(TLV::ContextTag(4), TLV::kTLVType_Structure, outerContainer));
+        ReturnErrorOnFailure(tlvWriter.Put(TLV::ContextTag(1), mIterationCount, true));
+        ReturnErrorOnFailure(tlvWriter.PutBytes(TLV::ContextTag(2), mSalt, mSaltLength));
+        ReturnErrorOnFailure(tlvWriter.EndContainer(outerContainer));
     }
 
-    resp->SetDataLength(static_cast<uint16_t>(resplen));
+    ReturnErrorOnFailure(tlvWriter.EndContainer(outerContainerType));
+    ReturnErrorOnFailure(tlvWriter.Finalize(&resp));
 
     // Update commissioning hash with the pbkdf2 param response that's being sent.
     ReturnErrorOnFailure(mCommissioningHash.AddData(ByteSpan{ resp->Start(), resp->DataLength() }));
     ReturnErrorOnFailure(SetupSpake2p(mIterationCount, ByteSpan(mSalt, mSaltLength)));
+
+    size_t sizeof_point = sizeof(mPoint);
     ReturnErrorOnFailure(mSpake2p.ComputeL(mPoint, &sizeof_point, mPASEVerifier.mL, kSpake2p_WS_Length));
 
-    mNextExpectedMsg = Protocols::SecureChannel::MsgType::PASE_Spake2p1;
+    mNextExpectedMsg = MsgType::PASE_Spake2p1;
```

Comment:
While we are here, suggest renaming the messages to PASE_Pake1..Pake3, since `Pake1`, `Pake2`, `Pake3` are the official names in the spec now.